bpf/kprobe_pwru.c

```diff
@@ -314,26 +442,49 @@ __set_tuple(struct tuple *tpl, void *data, u16 l3_off, bool is_ipv4) {
 		tpl->sport= BPF_CORE_READ(tcp, source);
 		tpl->dport= BPF_CORE_READ(tcp, dest);
 		bpf_probe_read_kernel(&tpl->tcp_flags, sizeof(tpl->tcp_flags), (void *)tcp + offsetof(struct tcphdr, window) - 1);
+		l4_off += sizeof(*tcp);
 	} else if (tpl->l4_proto == IPPROTO_UDP) {
 		struct udphdr *udp = (struct udphdr *) (data + l4_off);
 		tpl->sport= BPF_CORE_READ(udp, source);
 		tpl->dport= BPF_CORE_READ(udp, dest);
+		l4_off += sizeof(*udp);
 	}
+	return l4_off;
 }
 
-static __always_inline void
-set_tuple(struct sk_buff *skb, struct tuple *tpl) {
+static __always_inline s16
+set_tuple(struct sk_buff *skb, struct tuple *tpl, struct tuple *tunnel_tpl) {
 	void *skb_head = BPF_CORE_READ(skb, head);
 	u16 l3_off = BPF_CORE_READ(skb, network_header);
 
 	struct iphdr *l3_hdr = (struct iphdr *) (skb_head + l3_off);
 	u8 ip_vsn = BPF_CORE_READ_BITFIELD_PROBED(l3_hdr, version);
 
 	if (ip_vsn !=4 && ip_vsn != 6)
+		return -1;
+
+	bool is_ipv4 = ip_vsn == 4;
+	return __set_tuple(tpl, skb_head, l3_off, is_ipv4);
+}
+
+static __always_inline void
+set_tunnel(struct sk_buff *skb, struct tuple *tpl, struct tuple *tunnel_tpl, u16 l4_data_off) {
+	void *skb_head = BPF_CORE_READ(skb, head);
+
+	s16 tunnel_hdr_len = __tunnel_hdr_len(skb_head, tpl->sport, tpl->dport, l4_data_off);
+
+	if (tunnel_hdr_len == -1)
 		return;
 
+	struct ethhdr *inner = (struct ethhdr*) (skb_head + l4_data_off + tunnel_hdr_len);
```

Comment:
Couldn't we use `BPF_CORE_READ(skb, inner_mac_header);` ?
(don't recall if here the field already initialized)
(and below then use `BPF_CORE_READ(skb, inner_network_header);` etc)